wagtail/core/blocks/base.py

```diff
@@ -107,33 +112,56 @@ def set_name(self, name):
     def media(self):
         return forms.Media()
 
-    def html_declarations(self):
-        """
-        Return an HTML fragment to be rendered on the form page once per block definition -
-        as opposed to once per occurrence of the block. For example, the block definition
-            ListBlock(label="Shopping list", CharBlock(label="Product"))
-        needs to output a <script type="text/template"></script> block containing the HTML for
-        a 'product' text input, to that these can be dynamically added to the list. This
-        template block must only occur once in the page, even if there are multiple 'shopping list'
-        blocks on the page.
+    def get_layout(self):
+        return self.SIMPLE
 
-        Any element IDs used in this HTML fragment must begin with definition_prefix.
-        (More precisely, they must either be definition_prefix itself, or begin with definition_prefix
-        followed by a '-' character)
-        """
-        return ''
+    def prepare_value(self, value, errors=None):
+        return value
 
-    def js_initializer(self):
-        """
-        Returns a Javascript expression string, or None if this block does not require any
-        Javascript behaviour. This expression evaluates to an initializer function, a function that
-        takes the ID prefix and applies JS behaviour to the block instance with that value and prefix.
+    def prepare_for_react(self, parent_block, value,
+                          type_name=None, errors=None):
+        if type_name is None:
+            type_name = self.name
+        value = self.prepare_value(value, errors=errors)
+        if parent_block is None:
+            return value
+        return BlockData({
+            'id': str(uuid4()),
+            'type': type_name,
+            'hasError': bool(errors),
+            'value': value,
+        })
 
-        The parent block of this block (or the top-level page code) must ensure that this
-        expression is not evaluated more than once. (The resulting initializer function can and will be
-        called as many times as there are instances of this block, though.)
-        """
-        return None
+    def get_blocks_container_html(self, errors=None):
+        help_text = getattr(self.meta, 'help_text', None)
+        non_block_errors = get_non_block_errors(errors)
+        if help_text or non_block_errors:
+            return render_to_string(
+                'wagtailadmin/block_forms/blocks_container.html',
+                {
+                    'help_text': help_text,
+                    'non_block_errors': non_block_errors,
+                }
+            )
+
+    def get_definition(self):
+        definition = {
+            'key': self.name,
+            'label': capfirst(self.label),
+            'required': self.required,
+            'layout': self.get_layout(),
+            'dangerouslyRunInnerScripts': True,
```

Comment:
Actually we have to consider this carefully since there was a whole lot of work done in #4583 in order to eliminate all `eval`s for CSP purposes #1288.

Reply:
See https://github.com/thibaudcolas/react-streamfield/pull/1#discussion_r249226678 for a potential solution, with an implementation that's closer to what Wagtail does at the moment (I think).

Reply:
See in my answer how this doesn’t seem possible to do that without an `eval` in React: https://github.com/thibaudcolas/react-streamfield/pull/1/files#discussion_r258552925